### PR TITLE
Create Node Users Service

### PR DESCRIPTION
**Pull request by chamerling**
It also introduces some stuff on companies and prepare the work for twake-console export.

### CODE REVIEW

File: twake/backend/node/src/core/platform/services/database/services/orm/connectors/cassandra/query-builder.ts [hunk 1]
```diff
@@ -49,13 +49,16 @@ export function buildSelectQuery<Entity>(
...
     ...(buildComparison(findOptions) || []),
     ...(buildIn(findOptions) || []),
     ...(buildLike(findOptions) || []),
-  ].join(" AND ")}`;
+  ].join(" AND ")}`.trimEnd();
 
+  if (query.endsWith(" WHERE")) {
+    query = query.slice(0, -" WHERE".length)
+  }
   return query;
```
Comment (RomaricMourgues) [inline twake/backend/node/src/core/platform/services/database/services/orm/connectors/cassandra/query-builder.ts:62]: Why don't you create a variable where before and then decide to append it or not ?

```
const where = [ ...where,
    ...(buildComparison(findOptions) || []),	    ...(buildComparison(findOptions) || []),
    ...(buildIn(findOptions) || []),	    ...(buildIn(findOptions) || []),
    ...(buildLike(findOptions) || []),	    ...(buildLike(findOptions) || []),
  ].join(" AND ")}`;

const query = "SELECT * FROM ${options.keyspace}.${entityDefinition.name} " + (where.length ? (" WHERE "+where) : "")
```

File: twake/backend/node/src/services/user/entities/company_user.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+// backward compatibility with PHP where companies used to be `group`
+export const TYPE = "group_user";
+
+/**
+ * Link between a company and a user
+ */
+@Entity(TYPE, {
+  primaryKey: [["group_id"], ["user_id", "id"]],
+  type: TYPE,
+})
```
Comment (RomaricMourgues) [inline twake/backend/node/src/services/user/entities/company_user.ts:12]: Here we'll have an issue when we'll want to get all companies of one user. For now it is okay, but we'll need to create an other table or store companies id directly in user entity.

File: twake/backend/node/src/services/user/web/schemas.ts [hunk 3]
```diff
@@ -1,20 +1,28 @@
...
+    id: { type: "string" },
+    connected: { type: "boolean" },
+    creationdate: { type: "string" },
+    firstname: { type: "string" },
+    lastname: { type: "string" },
+    usernamecanonical: { type: "string" },
+    emailcanonical: { type: "string" },
+    phone: { type: "string" },
+    timezone: { type: "string" },
+    groups: { type: "array" },
```
Comment (RomaricMourgues) [inline twake/backend/node/src/services/user/web/schemas.ts:13]: Here we should use a better schema for the future, I will document this.
At least we could replace groups by companies ?

**Review by RomaricMourgues — CHANGES_REQUESTED** [review]
Small changes if you want, not necessary, rest seems good to me 👍

File: twake/backend/node/src/services/user/entities/company_user.ts (hunk 2 above)
Comment (RomaricMourgues) [reply twake/backend/node/src/services/user/entities/company_user.ts:12]: And by the way: `[["group_id"], ["user_id", "id"]],` => `[["group_id"], "user_id", "id"],`

File: twake/backend/node/src/services/user/web/schemas.ts (hunk 3 above)
Comment (chamerling) [reply twake/backend/node/src/services/user/web/schemas.ts:13]: Yes and no :)
I just want to map the current schema. When import will be OK and when we will start using the node backend to manage companies, we will have a new, cleaner schema, and cleaner fields, and cleaner parameter names

File: twake/backend/node/src/core/platform/services/database/services/orm/connectors/cassandra/query-builder.ts (hunk 1 above)
Comment (chamerling) [reply twake/backend/node/src/core/platform/services/database/services/orm/connectors/cassandra/query-builder.ts:62]: Why not :)

File: twake/backend/node/src/services/user/entities/company_user.ts (hunk 2 above)
Comment (chamerling) [reply twake/backend/node/src/services/user/entities/company_user.ts:12]: Yes!

**Review by RomaricMourgues — APPROVED** [review]

**Pull request merged**